std/encoding/csv.ts

```diff
@@ -49,70 +56,168 @@ function chkOptions(opt: ReadOptions): void {
+          if (line.startsWith(quote)) {
+            // `""` sequence (append quote).
+            recordBuffer += quote;
+            line = line.substring(quoteLen);
+          } else if (line.startsWith(opt.comma)) {
+            // `","` sequence (end of field).
+            line = line.substring(commaLen);
+            fieldIndexes.push(recordBuffer.length);
+            continue parseField;
+          } else if (lengthNL(line) === line.length) {
+            // `"\n` sequence (end of line).
+            fieldIndexes.push(recordBuffer.length);
+            break parseField;
+          } else if (opt.lazyQuotes) {
+            // `"` sequence (bare quote).
+            recordBuffer += quote;
+          } else {
+            // `"*` sequence (invalid non-escaped quote).
+            quoteError = ErrQuote;
+            break parseField;
+          }
+        } else if (line.length > 0 || !(await isEOF(tp))) {
+          // Hit end of line (copy all data so far).
+          recordBuffer += line;
+          const r = await readLine(tp);
+          if (r === Deno.EOF) {
+            if (!opt.lazyQuotes) {
+              quoteError = ErrQuote;
+              break parseField;
+            }
+            fieldIndexes.push(recordBuffer.length);
+            break parseField;
+          }
+          recordBuffer += "\n"; // preserve line feed (This is because TextProtoReader removes it.)
+          line = r;
+        } else {
+          // Abrupt end of file (EOF on error).
+          if (!opt.lazyQuotes) {
+            quoteError = ErrQuote;
+            break parseField;
+          }
+          fieldIndexes.push(recordBuffer.length);
+          break parseField;
+        }
       }
     }
-    return r;
-  });
+  }
   if (quoteError) {
-    throw new ParseError(Startline, lineIndex, 'bare " in non-quoted-field');
+    throw new ParseError(Startline, lineIndex, quoteError);
+  }
+  const result = [] as string[];
+  let preIdx = 0;
+  for (const i of fieldIndexes) {
+    result.push(recordBuffer.slice(preIdx, i));
+    preIdx = i;
   }
   return result;
 }
 
+async function isEOF(tp: TextProtoReader): Promise<boolean> {
+  return (await tp.r.peek(0)) === Deno.EOF;
+}
+
+async function readLine(tp: TextProtoReader): Promise<string | Deno.EOF> {
+  let line: string;
+  const r = await tp.readLine();
+  if (r === Deno.EOF) return Deno.EOF;
+  line = r;
+
+  // For backwards compatibility, drop trailing \r before EOF.
+  if ((await isEOF(tp)) && line.length > 0 && line[line.length - 1] === "\r") {
+    line = line.substring(0, line.length - 1);
+  }
+
+  // Normalize \r\n to \n on all input lines.
+  if (
+    line.length >= 2 &&
+    line[line.length - 2] === "\r" &&
+    line[line.length - 1] === "\n"
+  ) {
+    line = line.substring(0, line.length - 2);
+    line = line + "\n";
+  }
+
+  return line;
+}
+
+function lengthNL(s: string): number {
+  if (s.length > 0 && s[s.length - 1] === "\n") {
+    return 1;
+  }
+  return 0;
```

Comment:
s.endsWith(“\n”)

Maybe the function could be called endsWithNewLine

Reply:
I tought about it well.

In the original Go's csv module, the `lengthNL` function is used for counting trailing `\n`.

However, Deno's `TextProtoReader.readLine` returns a single line without trailing `\r\n` and `\n`.

So we can simply remove `lengthNL` function and simplify conditional checks as follows:

---

https://github.com/denoland/deno/pull/4520/files#diff-129e29636adf93e67856596174b2f486R136-R140

Before:

```typescript
          } else if (lengthNL(line) === line.length) {
            // `"\n` sequence (end of line).
            fieldIndexes.push(recordBuffer.length);
            break parseField;
          } else if (opt.lazyQuotes) {
```

After:

```typescript
          } else if (0 === line.length) {
            // `"\n` sequence (end of line).
            fieldIndexes.push(recordBuffer.length);
            break parseField;
          } else if (opt.lazyQuotes) {
```

Reply:
less code is better